src/lib/cache/memory/memory.go

```diff
@@ -117,27 +117,60 @@ func (c *Cache) Save(ctx context.Context, key string, value interface{}, expirat
 	return nil
 }
 
-// Keys returns the key matched by prefixes.
-func (c *Cache) Keys(ctx context.Context, prefixes ...string) ([]string, error) {
-	// if no prefix, means match all keys.
-	matchAll := len(prefixes) == 0
-	// range map to get all keys
-	keys := make([]string, 0)
-	c.storage.Range(func(k, v interface{}) bool {
-		ks := k.(string)
-		if matchAll {
-			keys = append(keys, ks)
-		} else {
-			for _, p := range prefixes {
-				if strings.HasPrefix(ks, c.opts.Key(p)) {
-					keys = append(keys, strings.TrimPrefix(ks, c.opts.Prefix))
+// Scan scans the keys matched by match string
+func (c *Cache) Scan(ctx context.Context, match string) (cache.Iterator, error) {
+	return &ScanIterator{storage: &c.storage, prefix: c.opts.Prefix, match: match}, nil
+}
+
+// ScanIterator is a ScanIterator for memory cache
+type ScanIterator struct {
+	once    sync.Once
+	storage *sync.Map
+	mu      sync.Mutex
+	pos     int
+	keys    []string
+	match   string
+	prefix  string
+}
+
+// Next checks whether has the next element
+func (i *ScanIterator) Next(ctx context.Context) bool {
+	i.once.Do(func() {
+		i.storage.Range(func(k, v interface{}) bool {
+			matched := true
+			if i.match != "" {
+				matched = strings.Contains(k.(string), i.match)
+			}
+
+			if matched {
+				if v.(*entry).isExpirated() {
+					i.storage.Delete(k)
+				} else {
+					i.keys = append(i.keys, strings.TrimPrefix(k.(string), i.prefix))
 				}
 			}
-		}
-		return true
+			return true
+		})
 	})
 
-	return keys, nil
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	i.pos++
+	return i.pos <= len(i.keys)
+}
+
+// Val returns the key
```

Comment:
It might be confusing, because the Val is not the key-value's value

Reply:
As the Scan command is only used for scan keys not values, so the Val() is return the value of key(just keep consistent with `go-redis` library).